### PR TITLE
chore: upgrade vllm from v0.9.0 to v0.9.2

### PR DESCRIPTION
**Pull request by yuki-666**
# What does this PR do ?
Upgrade vllm from v0.9.0 to v0.9.2 .

This could help:
1. NCCL issue when using non-colocated: https://github.com/NVIDIA-NeMo/RL/issues/722
2. gemma3 model support: https://github.com/NVIDIA-NeMo/RL/pull/676
3. VLM support

## Test Result
WIP

# Issues
Closes https://github.com/NVIDIA-NeMo/RL/issues/722.

### CODE REVIEW

**Review by terrykong — COMMENTED** [review]
Since v0.10.0 is out, are we able to jump straight to that one?

https://github.com/vllm-project/vllm/releases/tag/v0.10.0

**Comment by SahilJain314** [comment]
The release notes from 0.10.0 look good. Particularly this:
```
RLHF Support: new RPC methods for runtime weight reloading (https://github.com/vllm-project/vllm/pull/20096) and config updates (https://github.com/vllm-project/vllm/pull/20095), logprobs mode for selecting which stage of logprobs to return (https://github.com/vllm-project/vllm/pull/21398).
```

We need to be careful to select the post-processing ones for our case. The RPC methods should be useful as well.

**Comment by SahilJain314** [comment]
Blocking merging #543 until vllm is updated to 0.10.0

**Comment by SahilJain314** [comment]
raw logprobs are fine for now (same as v1). But are actually incorrect for top-k, top-p sampling (even though it _appears_ as though it would work correctly). Both the logprobs from vllm and the training framework are incorrect since they haven't been processed, which introduces error. (I believe the claim from #773 that v1 is fine is incorrect)

We have 2 options for how to mitigate this:
- use the post-processing from #773 on _both_ vllm raw_logprobs and megatron training
- use vllm post-processed logprobs and apply the processing to just megatron.

I'm leaning strongly towards the latter as it leaves less room for us to make mistakes and not notice. We should track this in #69.

**Comment by SahilJain314** [comment]
@yuki-666 I see in some of the plots that the new vllm version is much more 'spiky' in logprob error. This is fine (spikes are occasional) but would you mind also adding plots with the logprob error clamped to a max value of 1.2? I'd like to confirm that the typical value has not increased at all.

File: nemo_rl/models/generation/vllm.py [hunk 1]
```diff
@@ -325,7 +325,9 @@ def _patch_vllm_init_workers_ray():
         llm_kwargs = dict(
             model=self.model_name,
             load_format=load_format,
-            skip_tokenizer_init=self.cfg["vllm_cfg"]["skip_tokenizer_init"],
+            # vllm==0.10.0 breaks skip_tokenizer_init=True.
+            # This will be reverted to `self.cfg["vllm_cfg"]["skip_tokenizer_init"]` once https://github.com/vllm-project/vllm/issues/21846 is resolved.
+            skip_tokenizer_init=False,
```
Comment (SahilJain314) [inline nemo_rl/models/generation/vllm.py:330]: Can we create a bug so we don't forget this?

File: tests/unit/models/generation/test_vllm_generation.py [hunk 2]
```diff
@@ -1274,7 +1274,7 @@ async def test_vllm_refit_non_collocated_update_weights(
         print(f"Error during generation_cluster_separate shutdown: {e}")
 
 
-@pytest.mark.timeout(210)
+@pytest.mark.timeout(400)
```
Comment (SahilJain314) [inline tests/unit/models/generation/test_vllm_generation.py]: Why did this test get significantly slower here?

File: nemo_rl/models/generation/vllm.py (hunk 1 above)
Comment (yuki-97) [reply nemo_rl/models/generation/vllm.py:330]: https://github.com/NVIDIA-NeMo/RL/issues/818

File: tests/unit/models/generation/test_vllm_generation.py (hunk 2 above)
Comment (yuki-97) [reply tests/unit/models/generation/test_vllm_generation.py]: @yfw noticed that the vllm startup time was a little bit slower when testing `vllm==0.9.2`, and `vllm==0.10.0` has the same issue. so I have to increase the timeout.

```bash
209.02s call     tests/unit/models/generation/test_vllm_generation.py::test_vllm_generation_with_megatron_training[1]
113.36s call     tests/unit/models/generation/test_vllm_generation.py::test_vllm_generation_with_megatron_training[2]
```

**Comment by yuki-666** [comment]
> @yuki-666 I see in some of the plots that the new vllm version is much more 'spiky' in logprob error. This is fine (spikes are occasional) but would you mind also adding plots with the logprob error clamped to a max value of 1.2? I'd like to confirm that the typical value has not increased at all.

@SahilJain314  Added, seems fine, only the spikes > 1.05 (except the two VLMs), you can take a look.

**Comment by yuki-666** [comment]
> raw logprobs are fine for now (same as v1). But are actually incorrect for top-k, top-p sampling (even though it _appears_ as though it would work correctly). Both the logprobs from vllm and the training framework are incorrect since they haven't been processed, which introduces error. (I believe the claim from #773 that v1 is fine is incorrect)
> 
> We have 2 options for how to mitigate this:
> 
> * use the post-processing from [[1/2] Top-k and Top-p support for dtensor worker with vLLM V0 when `TP==1` #773](https://github.com/NVIDIA-NeMo/RL/pull/773) on _both_ vllm raw_logprobs and megatron training
> * use vllm post-processed logprobs and apply the processing to just megatron.
> 
> I'm leaning strongly towards the latter as it leaves less room for us to make mistakes and not notice. We should track this in #69.

I feel so too, although `raw_logprobs` is ok for temperature, the logits when using top-k/top-p will become incorrect if not actually applying top-k/top-p on them. And agreed with the latter one.

I think we can get this PR with `raw_logprobs` merged first. Then with @Dazz993 's help on https://github.com/NVIDIA-NeMo/RL/pull/773 and https://github.com/NVIDIA-NeMo/RL/pull/774, to support it at training side (both dtensor and megatron) and change `raw_logprobs` to `processed_logprobs` that time.

@SahilJain314  wdyt?

File: tests/unit/models/generation/test_vllm_generation.py (hunk 2 above)
Comment (yfw) [reply tests/unit/models/generation/test_vllm_generation.py]: For added context, I had noticed vllm startup time was a bit slower in the new version:

0.9.2:
(VllmGenerationWorker pid=2449082) INFO 07-17 10:56:24 [core.py:172] init engine (profile, create kv cache, warmup model) took 42.04 seconds

0.9.0:
(VllmGenerationWorker pid=2469150) INFO 07-17 10:59:48 [core.py:167] init engine (profile, create kv cache, warmup model) took 0.51 seconds

**Comment by terrykong** [comment]
Also want to just leave a comment that 0.10.0 version may fix this issue I've observed in gemma runs

https://github.com/vllm-project/vllm/pull/19449

**Review by rohitrango — APPROVED** [review]
